mongo/client_bulk_write.go

```diff
@@ -0,0 +1,703 @@
+			MarshalValueEncoderFn: newEncoderFn(mb.client.bsonOpts, mb.client.registry),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	var cursor *Cursor
+	cursor, err = newCursor(bCursor, mb.client.bsonOpts, mb.client.registry)
+	if err != nil {
+		return err
+	}
+	defer cursor.Close(ctx)
+
+	ok := true
+	for cursor.Next(ctx) {
+		var cur cursorInfo
+		err = cursor.Decode(&cur)
+		if err != nil {
+			return err
+		}
+		if int(cur.Idx) >= len(mb.cursorHandlers) {
+			continue
+		}
+		ok = mb.cursorHandlers[int(cur.Idx)](&cur, cursor.Current) && ok
+	}
+	err = cursor.Err()
+	if err != nil {
+		return err
+	}
+	if mb.ordered && (writeCmdErr.WriteConcernError != nil || !ok || !res.Ok || res.NErrors > 0) {
+		return ClientBulkWriteException{
+			WriteConcernErrors: mb.writeConcernErrors,
+			WriteErrors:        mb.writeErrors,
+			PartialResult:      mb.result,
+		}
+	}
+	return nil
+}
+
+func (mb *modelBatches) appendDeleteResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.DeleteResults == nil {
+			mb.result.DeleteResults = make(map[int]ClientBulkWriteDeleteResult)
+		}
+		mb.result.DeleteResults[idx] = ClientBulkWriteDeleteResult{int64(cur.N)}
+	}
+
+	return true
+}
+
+func (mb *modelBatches) appendInsertResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.InsertResults == nil {
+			mb.result.InsertResults = make(map[int]ClientBulkWriteInsertResult)
+		}
+		mb.result.InsertResults[idx] = ClientBulkWriteInsertResult{mb.newIDMap[idx]}
+	}
+
+	return true
+}
+
+func (mb *modelBatches) appendUpdateResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.UpdateResults == nil {
+			mb.result.UpdateResults = make(map[int]ClientBulkWriteUpdateResult)
+		}
+		result := ClientBulkWriteUpdateResult{
+			MatchedCount: int64(cur.N),
+		}
+		if cur.NModified != nil {
+			result.ModifiedCount = int64(*cur.NModified)
+		}
+		if cur.Upserted != nil {
+			result.UpsertedID = cur.Upserted.ID
+		}
+		mb.result.UpdateResults[idx] = result
+	}
+
+	return true
+}
+
+type clientInsertDoc struct {
+	namespace int
+	document  interface{}
+}
+
+func (d *clientInsertDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (interface{}, bsoncore.Document, error) {
```

Comment:
Is it possible to omit these duplicated types and define the `marshal` functions on the exported type instead? Do the duplicated types serve any additional purpose?